src/index.ts

```diff
@@ -1,3 +1,149 @@
-export const cat = {
-  say: 'meow',
-};
+import { SCALE_FACTOR, TILEGRID_WORLD_CRS84 } from './constants';
+import { BoundingBox, LonLat, Tile, TileGrid } from './interfaces';
+import { Zoom } from './types';
+import { validateBoundingBox, validateLonlat, validateMetatile, validateTile, validateTileGrid, validateZoomLevel } from './validations';
+
+function clamp(value: number, minValue: number, maxValue: number): number {
+  if (value < minValue) {
+    return minValue;
+  }
+
+  if (value > maxValue) {
+    return maxValue;
+  }
+
+  return value;
+}
+
+function tileHeight(zoom: Zoom, referenceTileGrid: TileGrid): number {
+  return (
+    (referenceTileGrid.boundingBox.north - referenceTileGrid.boundingBox.south) / (referenceTileGrid.numberOfMinLevelTilesY * SCALE_FACTOR ** zoom)
+  );
+}
+
+function tileWidth(zoom: Zoom, referenceTileGrid: TileGrid): number {
+  return (
+    (referenceTileGrid.boundingBox.east - referenceTileGrid.boundingBox.west) / (referenceTileGrid.numberOfMinLevelTilesX * SCALE_FACTOR ** zoom)
+  );
+}
+
+function* tilesGenerator(limits: [number, number, number, number], zoom: Zoom, metatile: number): Generator<Tile, undefined, undefined> {
+  for (let y = limits[1]; y <= limits[3]; y++) {
+    for (let x = limits[0]; x <= limits[2]; x++) {
+      yield { x: x, y: y, z: zoom, metatile: metatile };
+    }
+  }
+
+  return;
+}
+
+export function boundingBoxToTiles(
+  bbox: BoundingBox,
+  zoom: Zoom,
+  metatile = 1,
+  referenceTileGrid: TileGrid = TILEGRID_WORLD_CRS84
+): Generator<Tile, undefined, undefined> {
+  validateBoundingBox(bbox);
+  validateMetatile(metatile);
+  validateTileGrid(referenceTileGrid);
+  validateZoomLevel(zoom, referenceTileGrid);
+
+  const upperLeftTile = lonLatZoomToTile({ lon: bbox.west, lat: bbox.north }, zoom, metatile, referenceTileGrid);
+  const lowerRightTile = lonLatZoomToTile({ lon: bbox.east, lat: bbox.south }, zoom, metatile, referenceTileGrid);
+
+  return tilesGenerator([upperLeftTile.x, upperLeftTile.y, lowerRightTile.x, lowerRightTile.y], zoom, metatile);
+}
+
+export function zoomShift(zoom: Zoom, referenceTileGrid: TileGrid, targetTileGrid: TileGrid): Zoom {
+  validateTileGrid(referenceTileGrid);
+  validateTileGrid(targetTileGrid);
+  validateZoomLevel(zoom, referenceTileGrid);
+
+  const scale = referenceTileGrid.wellKnownScaleSet.scaleDenominators.get(zoom);
+  if (scale === undefined) {
+    // the value is validated before so this should be unreachable
+    throw new Error('zoom level is not part of the given well known scale set');
+  }
+
+  let matchingZoom: Zoom | undefined;
+  targetTileGrid.wellKnownScaleSet.scaleDenominators.forEach((targetScaleDenominator, targetZoom) => {
+    if (targetScaleDenominator === scale) {
+      matchingZoom = targetZoom;
+    }
+  });
```

Comment:
consider using for..of loop so you can return a result once it was found instead of looping all over.

Reply:
done